README
JRuby/LDAP
-----------

This is an interface compatible port of Ruby/LDAP - to allow binary LDAP usage through JRuby. This implementation is pure Ruby, but uses the Java Integration features of JRuby to access the JNDI libraries and through these implement the correct functionality.

Many classes are missing right now, but the base functionality should be there.


Usage
-----
require 'rubygems'
require 'ldap'

And then use it like all tutorials of Ruby/LDAP show you.


History.txt
== 0.0.2 (03/12/12)

- 1.9 compatibility (thanks Arkadiy Kraportov)

== 0.0.1 (10/09/07)

- Initial release.


test/setup.rb
require 'rubygems'
require 'bundler/setup'
require 'ldap'
require './username_and_password'

$LDAP_test_host = 'localhost'
$LDAP_test_port = LDAP::LDAP_PORT

def delete_tree(c, name)
  c.search(name, LDAP::LDAP_SCOPE_ONELEVEL, "(objectClass=*)", ['dn']) do |v|
    delete_tree c, v.get_dn
  end rescue nil
  c.delete(name)
end


c = LDAP::Conn.new($LDAP_test_host, $LDAP_test_port)
c.set_option(LDAP::LDAP_OPT_PROTOCOL_VERSION, 3)
c.bind($LDAP_test_username, $LDAP_test_password) do 
  # just clean out and create our tree

  delete_tree(c, 'o=ruby_ldap_test_tree') 
  
  c.add('o=ruby_ldap_test_tree', {'objectClass' => ['organization'], 'o' => ['ruby_ldap_test_tree']})
  c.add('o=sub_tree, o=ruby_ldap_test_tree', {'objectClass' => ['organization'], 'o' => ['sub_tree']})
end


module Test::Unit::Assertions
  def assert_exists_in_ldap dn, attrs = {}
    found = 0
    @conn.search(dn, LDAP::LDAP_SCOPE_BASE, "(objectClass=*)", attrs.keys) do |entry|
      found += 1

      assert_equal dn, entry.get_dn
      assert_equal attrs.keys.sort, entry.get_attributes.sort unless attrs == { }
      attrs.each do |k,v|
        assert_equal v, entry[k]
        assert_equal v, entry[k.downcase]
        assert_equal v, entry[k.upcase]
      end
    end
    
    assert_equal 1,found
  end

  def assert_dont_exists_in_ldap dn
    assert_raises(LDAP::ResultError) do 
      @conn.search(dn, LDAP::LDAP_SCOPE_BASE, "(objectClass=*)", []) { }
    end
  end
end


test/test_add.rb
require 'test/unit'
require 'setup'

class TestAdd < Test::Unit::TestCase
  def setup
    @conn = LDAP::Conn.new($LDAP_test_host, $LDAP_test_port)
    @conn.set_option(LDAP::LDAP_OPT_PROTOCOL_VERSION, 3)
    @conn.bind($LDAP_test_username, $LDAP_test_password)
  end
  
  def teardown
    @conn.unbind rescue nil
  end

  def test_simple_add
    @conn.add("o=fox3,o=ruby_ldap_test_tree", {'o' => ["fox3"], 'objectClass' => ['organization']})
    assert_exists_in_ldap "o=fox3,o=ruby_ldap_test_tree", {'o' => ["fox3"]}
  ensure 
    @conn.delete("o=fox3,o=ruby_ldap_test_tree") rescue nil
  end
end


test/test_ssl.rb
require 'test/unit'
require 'setup'

class TestSSLAdd# < Test::Unit::TestCase
  def setup
    @conn = LDAP::SSLConn.new($LDAP_test_host, LDAP::LDAPS_PORT)
    @conn.set_option(LDAP::LDAP_OPT_PROTOCOL_VERSION, 3)
    @conn.bind($LDAP_test_username, $LDAP_test_password)
  end
  
  def teardown
    @conn.unbind rescue nil
  end

  def test_simple_add
    @conn.add("o=fox3,o=ruby_ldap_test_tree", {'o' => ["fox3"], 'objectClass' => ['organization']})
    assert_exists_in_ldap "o=fox3,o=ruby_ldap_test_tree", {'o' => ["fox3"]}
  ensure 
    @conn.delete("o=fox3,o=ruby_ldap_test_tree") rescue nil
  end
end


lib/ldap.rb
require 'java'

module LDAP
  def self.err2string(err)
    case err||0
    when -1 then "Can't contact LDAP server"
    when 0 then "Success"
    when 1 then "Operations error"
    when 2 then "Protocol error"
    when 3 then "Time limit exceeded"
    when 4 then "Size limit exceeded"
    when 5 then "Compare False"
    when 6 then "Compare True"
    when 7 then "Authentication method not supported"
    when 8 then "Strong(er) authentication required"
    when 9 then "Partial results and referral received"
    when 10 then "Referral"
    when 11 then "Administrative limit exceeded"
    when 12 then "Critical extension is unavailable"
    when 13 then "Confidentiality required"
    when 14 then "SASL bind in progress"
    when 15 then "Unknown error"
    when 16 then "No such attribute"
    when 17 then "Undefined attribute type"
    when 18 then "Inappropriate matching"
    when 19 then "Constraint violation"
    when 20 then "Type or value exists"
    when 21 then "Invalid syntax"
    when 32 then "No such object"
    when 33 then "Alias problem"
    when 34 then "Invalid DN syntax"
    when 35 then "Entry is a leaf"
    when 36 then "Alias dereferencing problem"
    when 47 then "Proxy Authorization Failure"
    when 48 then "Inappropriate authentication"
    when 49 then "Invalid credentials"
    when 50 then "Insufficient access"
    when 51 then "Server is busy"
    when 52 then "Server is unavailable"
    when 53 then "Server is unwilling to perform"
    when 54 then "Loop detected"
    when 64 then "Naming violation"
    when 65 then "Object class violation"
    when 66 then "Operation not allowed on non-leaf"
    when 67 then "Operation not allowed on RDN"
    when 68 then "Already exists"
    when 69 then "Cannot modify object class"
    when 70 then "Results too large"
    when 71 then "Operation affects multiple DSAs"
    when 80 then "Internal (implementation specific) error"
    else "Unknown error"
    end
  end

  def self.load_configuration(attrs={})
    env = nil
    env = javax.naming.directory.InitialDirContext.new.environment rescue nil
    default = {'java.naming.factory.initial' => 'com.sun.jndi.ldap.LdapCtxFactory'}
    if env
      env2 = default.dup
      env.each do |k,v|
        env2[k.to_s] = v.to_s
      end
      env = env2
    else
      env = default.dup
    end
    env.merge! attrs
    @environment = env
  end
  
  def self.configuration(attrs = { })
    @environment.merge attrs
  end
end

require 'ldap/constants'
require 'ldap/conn'
require 'ldap/entry'
require 'ldap/error'
require 'ldap/mod'

LDAP.load_configuration


lib/ldap/mod.rb
module LDAP
  class << self
    def mod(mod_type, attr, vals)
      Mod.new(mod_type, attr, vals)
    end

    def hash2mods(mod_type, hash)
      hash.map do |key, value|
        mod(mod_type, key, value)
      end
    end
  end

  class Mod
    BasicAttributes = javax.naming.directory.BasicAttributes
    BasicAttribute = javax.naming.directory.BasicAttribute
    ModificationItem = javax.naming.directory.ModificationItem
    DirContext = javax.naming.directory.DirContext
    class << self
      def to_java_attributes(*attrs)
        attrs.inject(BasicAttributes.new) do |res, attr|
          res.put(attr.to_java_attribute)
          res
        end
      end

      def ruby_mod_to_java(mod_op)
        case (mod_op&~LDAP::LDAP_MOD_BVALUES)
          when LDAP::LDAP_MOD_ADD then DirContext::ADD_ATTRIBUTE
          when LDAP::LDAP_MOD_REPLACE then DirContext::REPLACE_ATTRIBUTE
          when LDAP::LDAP_MOD_DELETE then DirContext::REMOVE_ATTRIBUTE
          else raise LDAP::Error, "can't handle operation #{mod_op}"
        end
      end
      
      def to_java_modification_items(*attrs)
        attrs.map do |val|
          ModificationItem.new(ruby_mod_to_java(val.mod_op), val.to_java_attribute)
        end.to_java ModificationItem
      end
    end
    
    def initialize(mod_type, attr, vals)
      @type, @attr, @vals = mod_type, attr, vals
    end

    def mod_op #should be mod_type
      @type
    end

    def mod_type #should be attr
      @attr
    end

    def mod_vals
      @vals
    end
    
    def to_java_attribute
      v = BasicAttribute.new(self.mod_type)
      binary = mod_op & LDAP::LDAP_MOD_BVALUES
      if binary != 0
        self.mod_vals.each do |val|
          v.add(val.to_java_bytes)
        end
      else
        self.mod_vals.each do |val|
          v.add(val)
        end
      end
      v
    end
  end
end


lib/ldap/conn.rb
module LDAP  
  module ConnImplementation
    def controls(*args)
      raise "NOT IMPLEMENTED"
    end

    def get_option(*args)
      raise "NOT IMPLEMENTED"
    end
    
    # Modify the RDN of the entry with DN, +dn+, giving it the new RDN,
    # +new_rdn+. If +delete_old_rdn+ is *true*, the old RDN value will be deleted
    # from the entry.
    def modrdn(dn, new_rdn, delete_old_rdn)
      begin 
        if delete_old_rdn
          @context.rename(dn, new_rdn)
        else
          obj = @context.lookup(dn)
          @context.bind(new_rdn, obj)
        end
        @err = 0
      rescue javax.naming.NameAlreadyBoundException => e
        @err = 68
      rescue javax.naming.InvalidNameException => e
        @err = 34
      rescue javax.naming.NoPermissionException => e
        @err = 50
      rescue javax.naming.directory.SchemaViolationException => e
        @err = 65
      rescue javax.naming.NamingException => e
        @err = 21
      rescue javax.naming.NoPermissionException => e
        @err = 50
      rescue javax.naming.NamingException => e
        @err = -1
      end
      raise LDAP::ResultError.wrap(LDAP::err2string(@err), e) if @err != 0
      self
    end

    def perror(*args)
      raise "NOT IMPLEMENTED"
    end

    def referrals(*args)
      raise "NOT IMPLEMENTED"
    end
    
    def result2error(*args)
      raise "NOT IMPLEMENTED"
    end

    def __jndi_context
      @context
    end
    
    def initialize(host='localhost', port=LDAP_PORT)
      @host = host
      @port = port
    end
    
    def err
      @err || 0
    end
    
    def err2string(err)
      LDAP.err2string(err)
    end

    def simple_bind(dn=nil, password=nil, &block)
      bind(dn, password, LDAP_AUTH_SIMPLE, &block)
    end

    def bind(dn=nil, password=nil, method=LDAP_AUTH_SIMPLE)
      raise LDAP::Error, "already bound" if bound?

      url = @use_ssl ? "ldaps://#@host:#@port/" : "ldap://#@host:#@port/"
      base_env = {javax.naming.Context::PROVIDER_URL => url}
      base_env[javax.naming.Context::SECURITY_PRINCIPAL] = dn if dn
      base_env[javax.naming.Context::SECURITY_CREDENTIALS] = password if password

      @current_env = java.util.Hashtable.new(LDAP::configuration(base_env))

      begin 
        @context = javax.naming.directory.InitialDirContext.new(@current_env)
        @err = 0
      rescue javax.naming.NoPermissionException => e
        @err = 50
        raise LDAP::ResultError.wrap(LDAP::err2string(@err), e)
      rescue javax.naming.NamingException => e
        @err = -1
        raise LDAP::ResultError.wrap(LDAP::err2string(@err), e)
      end
      
      if !block_given?
        return self
      end

      begin 
        yield self

        return nil
      ensure
        unbind
      end
    end
    
    def set_option(opt, value)
      @err = 0
    end
    
    def add(dn, attrs)
      raise LDAP::InvalidDataError, "The LDAP handler has already unbound." unless bound?

      attrs = LDAP::hash2mods(LDAP::LDAP_MOD_ADD, attrs) if attrs.is_a?(Hash)

      begin 
        @context.create_subcontext(dn, LDAP::Mod.to_java_attributes(*attrs))
        @err = 0
      rescue javax.naming.NameNotFoundException => e
        @err = 32
        raise LDAP::ResultError.wrap(LDAP::err2string(@err), e)
      rescue javax.naming.InvalidNameException => e
        @err = 34
        raise LDAP::ResultError.wrap(LDAP::err2string(@err), e)
      rescue javax.naming.NoPermissionException => e
        @err = 50
        raise LDAP::ResultError.wrap(LDAP::err2string(@err), e)
      rescue javax.naming.directory.SchemaViolationException => e
        @err = 65
        raise LDAP::ResultError.wrap(LDAP::err2string(@err), e)
      rescue javax.naming.NamingException => e
        @err = 21
        raise LDAP::ResultError.wrap(LDAP::err2string(@err), e)
      end
      self
    end

    def modify(dn, attrs)
      raise LDAP::InvalidDataError, "The LDAP handler has already unbound." unless bound?

      attrs = LDAP::hash2mods(LDAP::LDAP_MOD_REPLACE, attrs) if attrs.is_a?(Hash)

      begin 
        @context.modify_attributes(dn, LDAP::Mod.to_java_modification_items(*attrs))
        @err = 0
      rescue javax.naming.NameNotFoundException => e
        @err = 32
        raise LDAP::ResultError.wrap(LDAP::err2string(@err), e)
      rescue javax.naming.InvalidNameException => e
        @err = 34
        raise LDAP::ResultError.wrap(LDAP::err2string(@err), e)
      rescue javax.naming.NoPermissionException => e
        @err = 50
        raise LDAP::ResultError.wrap(LDAP::err2string(@err), e)
      rescue javax.naming.directory.SchemaViolationException => e
        @err = 65
        raise LDAP::ResultError.wrap(LDAP::err2string(@err), e)
      rescue javax.naming.NamingException => e
        @err = 21
        raise LDAP::ResultError.wrap(LDAP::err2string(@err), e)
      end

      self
    end
    
    def delete(dn)
      raise LDAP::InvalidDataError, "The LDAP handler has already unbound." unless bound?

      begin
        @context.destroy_subcontext(dn)
        @err = 0
      rescue javax.naming.NameNotFoundException => e
        @err = 32
        raise LDAP::ResultError.wrap(LDAP::err2string(@err), e)
      rescue javax.naming.InvalidNameException => e
        @err = 34
        raise LDAP::ResultError.wrap(LDAP::err2string(@err), e)
      rescue javax.naming.NoPermissionException => e
        @err = 50
        raise LDAP::ResultError.wrap(LDAP::err2string(@err), e)
      rescue javax.naming.NamingException => e
        @err = 21
        raise LDAP::ResultError.wrap(LDAP::err2string(@err), e)
      end
      self
    end
    
    def search(base_dn, scope, filter, attrs=nil, attrsonly=nil, sec=0, usec=0, s_attr=nil, s_proc=nil)
      raise LDAP::InvalidDataError, "The LDAP handler has already unbound." unless bound?

      controls = javax.naming.directory.SearchControls.new
      controls.search_scope = scope

      if attrs && !attrs.empty?
        controls.returning_attributes = attrs.to_java(:string)
      end
      if attrsonly
        controls.returning_obj_flag = true
      end

      if sec != 0 || usec != 0
        controls.time_limit = usec/1000 + sec*1000
      end

      begin 
        @context.search(base_dn, filter, controls).each do |val|
          yield LDAP::Entry.create_from_search_result(val)
        end

        @err = 0
      rescue javax.naming.NameNotFoundException => e
        @err = 32
        raise LDAP::ResultError.wrap(LDAP::err2string(@err), e)
      rescue javax.naming.InvalidNameException => e
        @err = 34
        raise LDAP::ResultError.wrap(LDAP::err2string(@err), e)
      rescue javax.naming.NoPermissionException => e
        @err = 50
        raise LDAP::ResultError.wrap(LDAP::err2string(@err), e)
      end

      self
    end

    def search2(base_dn, scope, filter, attrs=nil, attrsonly=nil, sec=0, usec=0, s_attr=nil, s_proc=nil)
      arr = []
      search(base_dn, scope, filter, attrs, attrsonly, sec, usec, s_attr, s_proc) do |val|
        arr << LDAP::entry2hash(val)
      end
      arr
    end
    
    def unbind
      raise LDAP::InvalidDataError, "The LDAP handler has already unbound." unless bound?
      @context.close
      @err = 0
      @context = nil
    end
    
    def bound?
      !@context.nil?
    end
  end   
  

  class Conn
    class << self
      alias open new
    end

    def initialize(host='localhost', port=LDAP_PORT)
      super
      @use_ssl = false
    end
    
    include ConnImplementation
  end

  class SSLConn
    class << self
      alias open new
    end

    def initialize(host='localhost', port=LDAPS_PORT)
      super
      @use_ssl = true
    end
    
    include ConnImplementation
  end
end


lib/ldap/entry.rb

module LDAP
  class << self
    def entry2hash(entry)
      out = {}
      out['dn'] = [entry.dn]
      entry.get_attributes.each do |k|
        out[k] = entry[k]
      end
      out
    end
  end
  
  class Entry
    class << self
      def create_from_search_result(search_result)
        new(search_result.name_in_namespace, (search_result.attributes.get_all.inject({}){|hash, attr| 
                                                hash[attr.getID()] = attr.getAll.to_a; hash}))
      end
    end

    def initialize(dn, attributes)
      @dn = dn
      @attributes = attributes
      @keys = @attributes.keys
      @attributes = @keys.inject({}) do |hash, key|
        if @attributes[key].any?{|v| !v.is_a?(String)}
          @attributes[key] = @attributes[key].map do |val|
            val.is_a?(String) ? val : String.from_java_bytes(val)
          end
        end
        hash[key.downcase] = @attributes[key]
        hash
      end
    end
    
    def get_attributes
      @keys
    end

    alias attrs get_attributes

    def get_dn
      @dn
    end

    alias dn get_dn
    
    def get_values(attr)
      @attributes[attr.downcase]
    end
    
    alias vals get_values
    alias [] get_values
    
    def inspect
      super.split(' ').first + "\n#{(@keys.inject({}){|hash, name| hash[name] = @attributes[name.downcase];hash}.update('dn'=>[@dn])).inspect}>"
    end
    
    def hash
      @attributes.hash + @dn.hash
    end
  end
end


lib/ldap/constants.rb
module LDAP
  LDAPS_PORT = 636
  LDAP_ADMINLIMIT_EXCEEDED = 11
  LDAP_ALIAS_DEREF_PROBLEM = 36
  LDAP_ALIAS_PROBLEM = 33
  LDAP_ALREADY_EXISTS = 68
  LDAP_API_INFO_VERSION = 1
  LDAP_API_VERSION = 3001
  LDAP_AUTH_KRBV41 = 129
  LDAP_AUTH_KRBV42 = 130
  LDAP_AUTH_METHOD_NOT_SUPPORTED = 7
  LDAP_AUTH_NONE = 0
  LDAP_AUTH_SASL = 163
  LDAP_AUTH_SIMPLE = 128
  LDAP_AUTH_UNKNOWN = -6
  LDAP_BUSY = 51
  LDAP_COMPARE_FALSE = 5
  LDAP_COMPARE_TRUE = 6
  LDAP_CONFIDENTIALITY_REQUIRED = 13
  LDAP_CONSTRAINT_VIOLATION = 19
  LDAP_CONTROL_PAGEDRESULTS = "1.2.840.113556.1.4.319"
  LDAP_DECODING_ERROR = -4
  LDAP_DEREF_ALWAYS = 3
  LDAP_DEREF_FINDING = 2
  LDAP_DEREF_NEVER = 0
  LDAP_DEREF_SEARCHING = 1
  LDAP_ENCODING_ERROR = -3
  LDAP_FILTER_ERROR = -7
  LDAP_INAPPROPRIATE_AUTH = 48
  LDAP_INAPPROPRIATE_MATCHING = 18
  LDAP_INSUFFICIENT_ACCESS = 50
  LDAP_INVALID_CREDENTIALS = 49
  LDAP_INVALID_DN_SYNTAX = 34
  LDAP_INVALID_SYNTAX = 21
  LDAP_IS_LEAF = 35
  LDAP_LOCAL_ERROR = -2
  LDAP_LOOP_DETECT = 54
  LDAP_MOD_ADD = 0
  LDAP_MOD_BVALUES = 128
  LDAP_MOD_DELETE = 1
  LDAP_MOD_INCREMENT = 3
  LDAP_MOD_OP = 7
  LDAP_MOD_REPLACE = 2
  LDAP_NAMING_VIOLATION = 64
  LDAP_NOT_ALLOWED_ON_NONLEAF = 66
  LDAP_NOT_ALLOWED_ON_RDN = 67
  LDAP_NO_MEMORY = -10
  LDAP_NO_OBJECT_CLASS_MODS = 69
  LDAP_NO_SUCH_ATTRIBUTE = 16
  LDAP_NO_SUCH_OBJECT = 32
  LDAP_OBJECT_CLASS_VIOLATION = 65
  LDAP_OPERATIONS_ERROR = 1
  LDAP_OPT_API_FEATURE_INFO = 21
  LDAP_OPT_API_INFO = 0
  LDAP_OPT_CLIENT_CONTROLS = 19
  LDAP_OPT_DEREF = 2
  LDAP_OPT_DESC = 1
  LDAP_OPT_HOST_NAME = 48
  LDAP_OPT_OFF = 0
  LDAP_OPT_ON = 3117220
  LDAP_OPT_PROTOCOL_VERSION = 17
  LDAP_OPT_REFERRALS = 8
  LDAP_OPT_RESTART = 9
  LDAP_OPT_SERVER_CONTROLS = 18
  LDAP_OPT_SIZELIMIT = 3
  LDAP_OPT_TIMELIMIT = 4
  LDAP_OPT_X_SASL_AUTHCID = 24834
  LDAP_OPT_X_SASL_AUTHZID = 24835
  LDAP_OPT_X_SASL_MAXBUFSIZE = 24841
  LDAP_OPT_X_SASL_MECH = 24832
  LDAP_OPT_X_SASL_REALM = 24833
  LDAP_OPT_X_SASL_SECPROPS = 24838
  LDAP_OPT_X_SASL_SSF = 24836
  LDAP_OPT_X_SASL_SSF_EXTERNAL = 24837
  LDAP_OPT_X_SASL_SSF_MAX = 24840
  LDAP_OPT_X_SASL_SSF_MIN = 24839
  LDAP_OPT_X_TLS = 24576
  LDAP_OPT_X_TLS_ALLOW = 3
  LDAP_OPT_X_TLS_CACERTDIR = 24579
  LDAP_OPT_X_TLS_CACERTFILE = 24578
  LDAP_OPT_X_TLS_CERTFILE = 24580
  LDAP_OPT_X_TLS_CIPHER_SUITE = 24584
  LDAP_OPT_X_TLS_DEMAND = 2
  LDAP_OPT_X_TLS_HARD = 1
  LDAP_OPT_X_TLS_KEYFILE = 24581
  LDAP_OPT_X_TLS_NEVER = 0
  LDAP_OPT_X_TLS_RANDOM_FILE = 24585
  LDAP_OPT_X_TLS_REQUIRE_CERT = 24582
  LDAP_OPT_X_TLS_TRY = 4
  LDAP_OTHER = 80
  LDAP_PARAM_ERROR = -9
  LDAP_PARTIAL_RESULTS = 9
  LDAP_PORT = 389
  LDAP_PROTOCOL_ERROR = 2
  LDAP_REFERRAL = 10
  LDAP_RESULTS_TOO_LARGE = 70
  LDAP_SASL_BIND_IN_PROGRESS = 14
  LDAP_SASL_SIMPLE = nil
  LDAP_SCOPE_BASE = 0
  LDAP_SCOPE_ONELEVEL = 1
  LDAP_SCOPE_SUBTREE = 2
  LDAP_SERVER_DOWN = -1
  LDAP_SIZELIMIT_EXCEEDED = 4
  LDAP_STRONG_AUTH_NOT_SUPPORTED = 7
  LDAP_STRONG_AUTH_REQUIRED = 8
  LDAP_SUCCESS = 0
  LDAP_TIMELIMIT_EXCEEDED = 3
  LDAP_TIMEOUT = -5
  LDAP_TYPE_OR_VALUE_EXISTS = 20
  LDAP_UNAVAILABLE = 52
  LDAP_UNAVAILABLE_CRITICAL_EXTENSION = 12
  LDAP_UNDEFINED_TYPE = 17
  LDAP_UNWILLING_TO_PERFORM = 53
  LDAP_USER_CANCELLED = -8
  LDAP_VENDOR_NAME = "OpenLDAP"
  LDAP_VENDOR_VERSION = 20335
  LDAP_VERSION = 2
  LDAP_VERSION1 = 1
  LDAP_VERSION2 = 2
  LDAP_VERSION3 = 3
  LDAP_VERSION_MAX = 3
  MAJOR_VERSION = 0
  MINOR_VERSION = 9
  PATCH_VERSION = 7
  VERSION = "0.9.7"
end
